src/net/server/connection.rs

```diff
@@ -0,0 +1,633 @@
+    ///
+    /// Any requests received after the shutdown signal or requests still
+    /// in-flight will be abandoned.
+    ///
+    /// TODO: What does "abandoned" mean in practice here?
+    pub async fn run(mut self, command_rx: watch::Receiver<ServiceCommand>)
+    where
+        Svc::Single: Send,
+    {
+        self.metrics
+            .num_connections
+            .as_ref()
+            .unwrap()
+            .fetch_add(1, Ordering::Relaxed);
+
+        // TODO: Why use an Option and then Option::take()?
+        let stream = self.stream.take().unwrap();
+        self.run_until_error(command_rx, stream).await;
+
+        self.metrics
+            .num_connections
+            .as_ref()
+            .unwrap()
+            .fetch_sub(1, Ordering::Relaxed);
+    }
+}
+
+//--- Internal details
+
+impl<Stream, Buf, Svc> Connection<Stream, Buf, Svc>
+where
+    Stream: AsyncRead + AsyncWrite + Send + Sync + 'static,
+    Buf: BufSource + Send + Sync + 'static + Clone,
+    Buf::Output: Send + Sync + 'static,
+    Svc: Service<Buf::Output> + Send + Sync + 'static + Clone,
+{
+    async fn run_until_error(
+        &self,
+        mut command_rx: watch::Receiver<ServiceCommand>,
+        stream: Stream,
+    ) where
+        Svc::Single: Send,
+    {
+        let (mut stream_rx, stream_tx) = tokio::io::split(stream);
+        let (result_q_tx, mut result_q_rx) =
+            mpsc::channel::<CallResult<Svc::Target>>(10); // TODO: Take from configuration
+        let idle_timeout = chrono::Duration::seconds(3); // TODO: Take from configuration
+
+        let mut state =
+            StreamState::new(stream_tx, result_q_tx, idle_timeout);
+
+        let mut msg_size_buf = self
+            .buf_source
+            .create_sized(Message::<Buf::Output>::MIN_HDR_BYTES);
+
+        loop {
+            if let Err(err) = self
+                .transceive_one_request(
+                    &mut command_rx,
+                    &mut state,
+                    &mut stream_rx,
+                    &mut result_q_rx,
+                    &mut msg_size_buf,
+                )
+                .await
+            {
+                match err {
+                    ConnectionEvent::DisconnectWithoutFlush => {
+                        break;
+                    }
+                    ConnectionEvent::DisconnectWithFlush => {
+                        self.flush_write_queue(&mut state, &mut result_q_rx)
+                            .await;
+                        break;
+                    }
+                    ConnectionEvent::ReadSucceeded => {
+                        // A success is not an error, this shouldn't happen.
+                        unreachable!()
+                    }
+                    ConnectionEvent::ServiceError(err) => {
+                        error!("Service error: {}", err);
+                    }
+                }
+            }
+        }
+    }
+
+    async fn transceive_one_request(
```

Comment:
Yes we can clean this up in some way, I wanted the basic architecture reviewed first before spending any more time cleaning up.

Reply:
See commit https://github.com/NLnetLabs/domain/pull/274/commits/c4ae6d0ac601dc80bb5dc6e06e063a3563a74010 for changes I made relating to this feedback.

Reply:
Hmm, I think that commit is not cancel safe as the transceiver can now be interrupted part way through message receipt by a ready response message.

Reply:
Reworked. Though having tests for this would be good.

Reply:
`net/server/tests.rs` tests this now by checking that it gets the expected number of requests and responses, plus  `Connection::run_until_error()` will now panic in `#[cfg(test)]` compile mode if the transceiver (renamed to `DnsMessageReceiver` as it no longer handles responses since my earlier refactoring) was left in an incomplete read state.